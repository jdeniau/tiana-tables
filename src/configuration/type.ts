import { ConnectionObject } from '../component/Connection/types';

export type Configuration = {
  version: 1;
  theme: string;
  connections: Record<string, EncryptedConnectionObject>;
};

export type ConnectionAppState = {
  isActive: boolean; // TODO not handled yet
  activeDatabase: string;
  openedTable: string; // TODO not handled yet
};

export type EncryptedConnectionObject = {
  password: string;
  appState?: ConnectionAppState;
} & Omit<ConnectionObject, 'password'>;

export type EncryptedConfiguration = {
  connections: Record<string, EncryptedConnectionObject>;
} & Omit<Configuration, 'connections'>;
