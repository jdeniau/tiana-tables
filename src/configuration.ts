import { dialog, safeStorage } from 'electron';
import path from 'node:path';
import fs from 'node:fs';
import envPaths from 'env-paths';
import { ConnectionObject } from './component/Connection';

export type Configuration = {
  version: 1;
  connections: Record<string, EncryptedConnectionObject>;
};

type EncryptedConnectionObject = {
  password: string;
} & Omit<ConnectionObject, 'password'>;

type EncryptedConfiguration = {
  connections: Record<string, EncryptedConnectionObject>;
} & Omit<Configuration, 'connections'>;

const envPath = envPaths('FuzzyPotato', { suffix: '' });
const dataFilePath = path.resolve(envPath.config, 'config.json');

function encryptConnection(
  connection: ConnectionObject
): EncryptedConnectionObject {
  return {
    ...connection,
    password: safeStorage.encryptString(connection.password).toString('base64'),
  };
}

function decryptConnection(
  connection: EncryptedConnectionObject
): ConnectionObject {
  return {
    ...connection,
    password: safeStorage.decryptString(
      Buffer.from(connection.password, 'base64')
    ),
  };
}

export function readConfigurationFile(): null | Configuration {
  if (!fs.existsSync(dataFilePath)) {
    return null;
  }

  const dataString = fs.readFileSync(dataFilePath, 'utf-8');

  let config = JSON.parse(dataString) as EncryptedConfiguration;

  return {
    ...config,
    connections: Object.fromEntries(
      Object.entries(config.connections).map(([name, connection]) => [
        name,
        decryptConnection(connection),
      ])
    ),
  };
}
export function addConnectionToConfig(
  event: Electron.IpcMainInvokeEvent,
  name: string,
  connection: ConnectionObject
): void {
  let config = readConfigurationFile();

  if (!config) {
    config = {
      version: 1,
      connections: {},
    };
  }

  if (!config.connections) {
    config.connections = {};
  }

  config.connections[name] = encryptConnection(connection);

  fs.mkdirSync(envPath.config, { recursive: true });
  fs.writeFile(
    dataFilePath,
    JSON.stringify(config, null, 2),
    'utf-8',
    (err) => {
      if (err) {
        dialog.showErrorBox('Error', err.message);
      }
    }
  );
}
