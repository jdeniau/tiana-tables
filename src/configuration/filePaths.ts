import { app } from 'electron';
import { existsSync, mkdirSync } from 'node:fs';
import path, { dirname, resolve } from 'node:path';
import { isDevApp } from '../main-process/helpers';

export function getConfigurationPath() {
  const configDir = resolve(app.getPath('userData'), 'config');
  const configPath = resolve(configDir, 'config.json');

  return configPath;
}

export function getConfigurationFolder() {
  return dirname(getConfigurationPath());
}

export function createConfigurationFolderIfNotExists() {
  const configurationPah = getConfigurationPath();

  if (!existsSync(dirname(configurationPah))) {
    mkdirSync(dirname(configurationPah), { recursive: true });
  }
}

export function getLogPath() {
  return path.join(
    app.getPath('userData'),
    'logs',
    isDevApp() ? 'dev.log' : 'main.log'
  );
}

export function getLogFolder() {
  return dirname(getLogPath());
}
