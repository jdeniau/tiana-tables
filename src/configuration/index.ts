import { dialog, safeStorage } from 'electron';
import { resolve } from 'node:path';
import { existsSync, mkdirSync, readFileSync, writeFile } from 'node:fs';
import envPaths from 'env-paths';
import { ConnectionObject } from '../component/Connection/types';
import { DEFAULT_THEME } from '../theme';
import {
  Configuration,
  EncryptedConnectionObject,
  EncryptedConfiguration,
  ConnectionAppState,
} from './type';

// TODO use app.getPath('userData') to store the configuration file instead of env-paths
const envPath = envPaths('TianaTables', { suffix: '' });
const dataFilePath = resolve(envPath.config, 'config.json');

console.log('Configuration file path:', dataFilePath);

function getBaseConfig(): Configuration {
  return {
    version: 1,
    theme: DEFAULT_THEME.name,
    connections: {},
  };
}

function encryptConnection(
  connection: ConnectionObject
): EncryptedConnectionObject {
  return {
    ...connection,
    password: safeStorage.encryptString(connection.password).toString('base64'),
  };
}

function decryptConnection(
  connection: EncryptedConnectionObject
): ConnectionObject {
  return {
    ...connection,
    password: safeStorage.decryptString(
      Buffer.from(connection.password, 'base64')
    ),
  };
}

export function getConfiguration(): Configuration {
  if (!existsSync(dataFilePath)) {
    return getBaseConfig();
  }

  const dataString = readFileSync(dataFilePath, 'utf-8');

  if (!dataString) {
    return getBaseConfig();
  }

  const config = JSON.parse(dataString) as EncryptedConfiguration;

  return {
    ...config,
    connections: Object.fromEntries(
      Object.entries(config.connections ?? {}).map(([name, connection]) => [
        name,
        decryptConnection(connection),
      ])
    ),
  };
}

function writeConfiguration(config: Configuration): void {
  const encryptedConfig = {
    ...config,
    connections: Object.fromEntries(
      Object.entries(config.connections).map(([name, connection]) => [
        name,
        encryptConnection(connection),
      ])
    ),
  };

  mkdirSync(envPath.config, { recursive: true });
  writeFile(
    dataFilePath,
    JSON.stringify(encryptedConfig, null, 2),
    'utf-8',
    (err) => {
      if (err) {
        dialog.showErrorBox('Error', err.message);
      }
    }
  );
}

export function addConnectionToConfig(connection: ConnectionObject): void {
  const config = getConfiguration() ?? getBaseConfig();

  if (!config.connections) {
    config.connections = {};
  }

  config.connections[connection.name] = connection;

  writeConfiguration(config);
}

export function changeTheme(theme: string): void {
  const config = getConfiguration() ?? getBaseConfig();
  config.theme = theme;

  writeConfiguration(config);
}

export function updateConnectionState<K extends keyof ConnectionAppState>(
  connectionName: string,
  key: K,
  value: ConnectionAppState[K]
) {
  const config = getConfiguration();

  const connection = config.connections[connectionName];

  if (!connection) {
    return;
  }

  if (!connection.appState) {
    connection.appState = {
      isActive: false,
      activeDatabase: '',
      openedTable: '',
    };
  }

  connection.appState[key] = value;

  writeConfiguration(config);
}

export const testables = { getBaseConfig };
