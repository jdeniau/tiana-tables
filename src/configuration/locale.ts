export const DEFAULT_LOCALE = 'en';

export const LOCALE_LIST = ['en', 'fr'];
