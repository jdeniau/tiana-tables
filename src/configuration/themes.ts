import * as dracula from '../renderer/theme/dracula.json';
import * as visualStudio from '../renderer/theme/visualStudio.json';
import * as active4d from '../renderer/theme/active4d.json';

export interface TmTheme {
  readonly name: string;
  readonly author?: string;
  readonly settings: TmThemeSetting[];
}

const THEME_LIST_AS_ARRAY: TmTheme[] = [
  dracula,
  visualStudio,
  active4d,
] as const;

const DARK_THEME_LIST_NAME = [dracula.name];

type Theme = (typeof THEME_LIST_AS_ARRAY)[number];

export const THEME_LIST: Record<string, Theme> = {};
THEME_LIST_AS_ARRAY.forEach((t) => {
  THEME_LIST[t.name] = t;
});

export const DEFAULT_THEME = THEME_LIST_AS_ARRAY[0];

export function isDarkTheme(theme: Theme): boolean {
  return DARK_THEME_LIST_NAME.includes(theme.name);
}

export interface TmThemeSetting {
  readonly scope?: string | string[];
  readonly settings: Record<string, string | undefined>;
}

export interface TmThemeScopedSetting extends TmThemeSetting {
  readonly scope: string | string[];
}

export interface TmThemeGlobalSetting extends TmThemeSetting {
  readonly scope: undefined;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function isScopedSetting(o: { scope?: any; settings?: any }): boolean {
  return o.scope && o.settings;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function isUnscopedSetting(o: { scope?: any; settings?: any }): boolean {
  return !o.scope && o.settings;
}
