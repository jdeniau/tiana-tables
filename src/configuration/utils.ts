export function slugify(name: string): string {
  return name
    .toLowerCase()
    .replace(/ /g, '-')
    .replace(/[^a-z0-9-]/g, '')
    .replace(/-{1,}/g, '-')
    .replace(/-$/, '')
    .replace(/^-/, '');
}
