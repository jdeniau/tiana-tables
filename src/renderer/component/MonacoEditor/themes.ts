import { editor } from 'monaco-editor';
import {
  TmTheme,
  isDarkTheme,
  isUnscopedSetting,
} from '../../../configuration/themes';

// Taken from https://github.com/brijeshb42/monaco-themes/blob/2bf2a1f3a3a1e43e0b91c4a67376dd4051b5055d/src/index.js
// but monaco-themes needs a pList file to work
// It alse guess if the theme is dark or light, but we already have that information

function parseColor(color: string): string | null {
  if (!color.length) {
    return null;
  }

  if (color.length == 4) {
    color = color.replace(/[a-fA-F\d]/g, '$&$&');
  }

  if (color.length == 7) {
    return color;
  }

  if (color.length == 9) {
    return color; // substr(0, 7);
  } else {
    if (!color.match(/^#(..)(..)(..)(..)$/)) {
      throw new Error(`can't parse color ${color}`);
    }

    const rgba: Array<number> | undefined = color
      .match(/^#(..)(..)(..)(..)$/)
      ?.slice(1)
      .map(function (c) {
        return parseInt(c, 16);
      });

    if (!rgba) {
      throw new Error(`can't parse color ${color}`);
    }

    // @ts-expect-error string / number assignation issue, but will be joined the line after
    rgba[3] = (rgba[3] / 0xff).toPrecision(2);

    return 'rgba(' + rgba.join(', ') + ')';
  }
}

/* Mapped from vscode/src/vs/workbench/services/themes/electron-browser/themeCompatibility.ts */
const COLOR_MAP = [
  {
    tm: 'foreground',
    mn: 'editor.foreground',
  },
  {
    tm: 'background',
    mn: 'editor.background',
  },
  // {
  //   tm: 'foreground',
  //   mn: 'editorSuggestWidget.foreground',
  // },
  // {
  //   tm: 'background',
  //   mn: 'editorSuggestWidget.background',
  // },
  {
    tm: 'selection',
    mn: 'editor.selectionBackground',
  },
  {
    tm: 'inactiveSelection',
    mn: 'editor.inactiveSelectionBackground',
  },
  {
    tm: 'selectionHighlightColor',
    mn: 'editor.selectionHighlightBackground',
  },
  {
    tm: 'findMatchHighlight',
    mn: 'editor.findMatchHighlightBackground',
  },
  {
    tm: 'currentFindMatchHighlight',
    mn: 'editor.findMatchBackground',
  },
  {
    tm: 'hoverHighlight',
    mn: 'editor.hoverHighlightBackground',
  },
  {
    tm: 'wordHighlight',
    mn: 'editor.wordHighlightBackground',
  },
  {
    tm: 'wordHighlightStrong',
    mn: 'editor.wordHighlightStrongBackground',
  },
  {
    tm: 'findRangeHighlight',
    mn: 'editor.findRangeHighlightBackground',
  },
  {
    tm: 'findMatchHighlight',
    mn: 'peekViewResult.matchHighlightBackground',
  },
  {
    tm: 'referenceHighlight',
    mn: 'peekViewEditor.matchHighlightBackground',
  },
  {
    tm: 'lineHighlight',
    mn: 'editor.lineHighlightBackground',
  },
  {
    tm: 'rangeHighlight',
    mn: 'editor.rangeHighlightBackground',
  },
  {
    tm: 'caret',
    mn: 'editorCursor.foreground',
  },
  {
    tm: 'invisibles',
    mn: 'editorWhitespace.foreground',
  },
  {
    tm: 'guide',
    mn: 'editorIndentGuide.background',
  },
  {
    tm: 'activeGuide',
    mn: 'editorIndentGuide.activeBackground',
  },
  {
    tm: 'selectionBorder',
    mn: 'editor.selectionHighlightBorder',
  },
];

const ansiColorMap = [
  'ansiBlack',
  'ansiRed',
  'ansiGreen',
  'ansiYellow',
  'ansiBlue',
  'ansiMagenta',
  'ansiCyan',
  'ansiWhite',
  'ansiBrightBlack',
  'ansiBrightRed',
  'ansiBrightGreen',
  'ansiBrightYellow',
  'ansiBrightBlue',
  'ansiBrightMagenta',
  'ansiBrightCyan',
  'ansiBrightWhite',
];

ansiColorMap.forEach((color) => {
  COLOR_MAP.push({
    tm: color,
    mn: 'terminal.' + color,
  });
});
export function convertTextmateThemeToMonaco(
  rawData: TmTheme
): editor.IStandaloneThemeData {
  const firstSettings = rawData.settings[0];

  if (!isUnscopedSetting(firstSettings)) {
    throw new Error('Expected global settings in first position');
  }
  const globalSettings = firstSettings.settings;

  const rules: editor.ITokenThemeRule[] = [];

  rawData.settings.forEach((setting) => {
    if (!setting.settings) {
      return;
    }

    let scopes;

    if (typeof setting.scope === 'string') {
      scopes = setting.scope
        .replace(/^[,]+/, '')
        .replace(/[,]+$/, '')
        .split(',');
    } else if (Array.isArray(setting.scope)) {
      scopes = setting.scope;
    } else {
      scopes = [''];
    }

    const rule: Partial<editor.ITokenThemeRule> = {};
    const settings = setting.settings;

    if (settings.foreground) {
      const parsed = parseColor(settings.foreground);
      if (parsed) {
        rule.foreground = parsed.toLowerCase().replace('#', '');
      }
    }

    if (settings.background) {
      const parsed = parseColor(settings.background);
      if (parsed) {
        rule.background = parsed.toLowerCase().replace('#', '');
      }
    }

    if (settings.fontStyle && typeof settings.fontStyle === 'string') {
      rule.fontStyle = settings.fontStyle;
    }

    scopes.forEach((scope) => {
      if (!scope || !Object.keys(rule).length) {
        return;
      }
      const r = Object.assign({}, rule, {
        token: scope.trim(),
      });
      rules.push(r);
    });
  });

  const globalColors: editor.IColors = {};

  /* More properties to be added */
  COLOR_MAP.forEach((obj) => {
    const tmGlobalSetting = globalSettings[obj.tm];

    if (tmGlobalSetting) {
      const parsed = parseColor(tmGlobalSetting);

      if (parsed) {
        globalColors[obj.mn] = parsed;
      }
    }
  });

  const base = isDarkTheme(rawData) ? 'vs-dark' : 'vs';

  return {
    base,
    inherit: true,
    rules: rules,
    colors: globalColors,
  };
}
