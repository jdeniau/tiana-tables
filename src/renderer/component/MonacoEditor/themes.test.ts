import { describe, expect, test } from 'vitest';
import { THEME_LIST } from '../../../configuration/themes';
import { convertTextmateThemeToMonaco } from './themes';

describe('convertTextmateThemeToMonaco', () => {
  test.each(Object.keys(THEME_LIST))(
    'should convert %s theme to Monaco theme',
    (themeName) => {
      const theme = THEME_LIST[themeName];
      const monacoTheme = convertTextmateThemeToMonaco(theme);
      expect(monacoTheme).toMatchSnapshot();
    }
  );
});
