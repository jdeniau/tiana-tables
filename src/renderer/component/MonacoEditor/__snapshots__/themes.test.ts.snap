// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`convertTextmateThemeToMonaco > should convert Active4D theme to Monaco theme 1`] = `
{
  "base": "vs",
  "colors": {
    "editor.background": "#FFFFFF",
    "editor.foreground": "#3B3B3B",
    "editor.lineHighlightBackground": "#00000012",
    "editor.selectionBackground": "#BAD6FD",
    "editorCursor.foreground": "#000000",
    "editorWhitespace.foreground": "#BFBFBF",
  },
  "inherit": true,
  "rules": [
    {
      "background": "e2e9ff5e",
      "token": "text.html source.active4d",
    },
    {
      "foreground": "000000",
      "token": "text.xml",
    },
    {
      "foreground": "af82d4",
      "token": "comment.line",
    },
    {
      "foreground": "af82d4",
      "token": "comment.block",
    },
    {
      "foreground": "666666",
      "token": "string",
    },
    {
      "fontStyle": "bold",
      "foreground": "66ccff",
      "token": "string.interpolated variable",
    },
    {
      "foreground": "a8017e",
      "token": "constant.numeric",
    },
    {
      "fontStyle": "bold",
      "foreground": "66ccff",
      "token": "constant.other.date",
    },
    {
      "fontStyle": "bold",
      "foreground": "66ccff",
      "token": "constant.other.time",
    },
    {
      "foreground": "a535ae",
      "token": "constant.language",
    },
    {
      "fontStyle": "bold",
      "foreground": "6392ff",
      "token": "variable.other.local",
    },
    {
      "fontStyle": "bold",
      "foreground": "0053ff",
      "token": "variable",
    },
    {
      "foreground": "6988ae",
      "token": "variable.other.table-field",
    },
    {
      "fontStyle": "bold",
      "foreground": "006699",
      "token": "keyword",
    },
    {
      "foreground": "ff5600",
      "token": "storage",
    },
    {
      "foreground": "21439c",
      "token": "entity.name.type",
    },
    {
      "foreground": "21439c",
      "token": "entity.name.function",
    },
    {
      "foreground": "7a7a7a",
      "token": "meta.tag",
    },
    {
      "foreground": "016cff",
      "token": "entity.name.tag",
    },
    {
      "foreground": "963dff",
      "token": "entity.other.attribute-name",
    },
    {
      "fontStyle": "bold",
      "foreground": "45ae34",
      "token": "support.function",
    },
    {
      "foreground": "b7734c",
      "token": "support.constant",
    },
    {
      "foreground": "a535ae",
      "token": "support.type",
    },
    {
      "foreground": "a535ae",
      "token": "support.class",
    },
    {
      "foreground": "a535ae",
      "token": "support.variable",
    },
    {
      "background": "990000",
      "foreground": "ffffff",
      "token": "invalid",
    },
    {
      "background": "656565",
      "foreground": "ffffff",
      "token": "meta.diff",
    },
    {
      "background": "1b63ff",
      "foreground": "ffffff",
      "token": "meta.diff.range",
    },
    {
      "background": "ff7880",
      "foreground": "000000",
      "token": "markup.deleted.diff",
    },
    {
      "background": "98ff9a",
      "foreground": "000000",
      "token": "markup.inserted.diff",
    },
    {
      "foreground": "5e5e5e",
      "token": "source.diff",
    },
  ],
}
`;

exports[`convertTextmateThemeToMonaco > should convert Dracula theme to Monaco theme 1`] = `
{
  "base": "vs-dark",
  "colors": {
    "editor.background": "#282a36",
    "editor.foreground": "#f8f8f2",
    "editor.lineHighlightBackground": "#44475a",
    "editor.selectionBackground": "#44475a",
    "editor.selectionHighlightBorder": "#222218",
    "editorCursor.foreground": "#f8f8f0",
    "editorIndentGuide.activeBackground": "#9D550FB0",
    "editorWhitespace.foreground": "#3B3A32",
  },
  "inherit": true,
  "rules": [
    {
      "foreground": "6272a4",
      "token": "comment",
    },
    {
      "foreground": "f1fa8c",
      "token": "string",
    },
    {
      "foreground": "bd93f9",
      "token": "constant.numeric",
    },
    {
      "foreground": "bd93f9",
      "token": "constant.language",
    },
    {
      "foreground": "bd93f9",
      "token": "constant.character",
    },
    {
      "foreground": "bd93f9",
      "token": "constant.other",
    },
    {
      "foreground": "ffb86c",
      "token": "variable.other.readwrite.instance",
    },
    {
      "foreground": "ff79c6",
      "token": "constant.character.escaped",
    },
    {
      "foreground": "ff79c6",
      "token": "constant.character.escape",
    },
    {
      "foreground": "ff79c6",
      "token": "string source",
    },
    {
      "foreground": "ff79c6",
      "token": "string source.ruby",
    },
    {
      "foreground": "ff79c6",
      "token": "keyword",
    },
    {
      "foreground": "ff79c6",
      "token": "storage",
    },
    {
      "fontStyle": "italic",
      "foreground": "8be9fd",
      "token": "storage.type",
    },
    {
      "fontStyle": "underline",
      "foreground": "50fa7b",
      "token": "entity.name.class",
    },
    {
      "fontStyle": "italic underline",
      "foreground": "50fa7b",
      "token": "entity.other.inherited-class",
    },
    {
      "foreground": "50fa7b",
      "token": "entity.name.function",
    },
    {
      "fontStyle": "italic",
      "foreground": "ffb86c",
      "token": "variable.parameter",
    },
    {
      "foreground": "ff79c6",
      "token": "entity.name.tag",
    },
    {
      "foreground": "50fa7b",
      "token": "entity.other.attribute-name",
    },
    {
      "foreground": "8be9fd",
      "token": "support.function",
    },
    {
      "foreground": "6be5fd",
      "token": "support.constant",
    },
    {
      "fontStyle": " italic",
      "foreground": "66d9ef",
      "token": "support.type",
    },
    {
      "fontStyle": " italic",
      "foreground": "66d9ef",
      "token": "support.class",
    },
    {
      "background": "ff79c6",
      "foreground": "f8f8f0",
      "token": "invalid",
    },
    {
      "background": "bd93f9",
      "foreground": "f8f8f0",
      "token": "invalid.deprecated",
    },
    {
      "foreground": "cfcfc2",
      "token": "meta.structure.dictionary.json string.quoted.double.json",
    },
    {
      "foreground": "6272a4",
      "token": "meta.diff",
    },
    {
      "foreground": "6272a4",
      "token": "meta.diff.header",
    },
    {
      "foreground": "ff79c6",
      "token": "markup.deleted",
    },
    {
      "foreground": "50fa7b",
      "token": "markup.inserted",
    },
    {
      "foreground": "e6db74",
      "token": "markup.changed",
    },
    {
      "foreground": "bd93f9",
      "token": "constant.numeric.line-number.find-in-files - match",
    },
    {
      "foreground": "e6db74",
      "token": "entity.name.filename",
    },
    {
      "foreground": "f83333",
      "token": "message.error",
    },
    {
      "foreground": "eeeeee",
      "token": "punctuation.definition.string.begin.json - meta.structure.dictionary.value.json",
    },
    {
      "foreground": "eeeeee",
      "token": "punctuation.definition.string.end.json - meta.structure.dictionary.value.json",
    },
    {
      "foreground": "8be9fd",
      "token": "meta.structure.dictionary.json string.quoted.double.json",
    },
    {
      "foreground": "f1fa8c",
      "token": "meta.structure.dictionary.value.json string.quoted.double.json",
    },
    {
      "foreground": "50fa7b",
      "token": "meta meta meta meta meta meta meta.structure.dictionary.value string",
    },
    {
      "foreground": "ffb86c",
      "token": "meta meta meta meta meta meta.structure.dictionary.value string",
    },
    {
      "foreground": "ff79c6",
      "token": "meta meta meta meta meta.structure.dictionary.value string",
    },
    {
      "foreground": "bd93f9",
      "token": "meta meta meta meta.structure.dictionary.value string",
    },
    {
      "foreground": "50fa7b",
      "token": "meta meta meta.structure.dictionary.value string",
    },
    {
      "foreground": "ffb86c",
      "token": "meta meta.structure.dictionary.value string",
    },
  ],
}
`;

exports[`convertTextmateThemeToMonaco > should convert Visual Studio theme to Monaco theme 1`] = `
{
  "base": "vs",
  "colors": {
    "editor.background": "#FFFFFF",
    "editor.foreground": "#000000",
    "editor.lineHighlightBackground": "#00000012",
    "editor.selectionBackground": "#9DA7C3",
    "editorCursor.foreground": "#000000",
    "editorWhitespace.foreground": "#BFBFBF",
  },
  "inherit": true,
  "rules": [
    {
      "background": "2f8963",
      "foreground": "eefff8",
      "token": "markup.inserted.diff",
    },
    {
      "background": "2f8963",
      "foreground": "eefff8",
      "token": "meta.diff.header.to-file",
    },
    {
      "background": "c86a44",
      "foreground": "fff3ee",
      "token": "markup.deleted.diff",
    },
    {
      "background": "c86a44",
      "foreground": "fff3ee",
      "token": "meta.diff.header.from-file",
    },
    {
      "foreground": "008000",
      "token": "comment",
    },
    {
      "foreground": "000000",
      "token": "variable",
    },
    {
      "foreground": "0000ff",
      "token": "keyword",
    },
    {
      "foreground": "a31515",
      "token": "constant.numeric",
    },
    {
      "foreground": "000000",
      "token": "constant",
    },
    {
      "foreground": "000000",
      "token": "constant.language",
    },
    {
      "foreground": "0000ff",
      "token": "constant.language.boolean",
    },
    {
      "foreground": "a31515",
      "token": "string",
    },
    {
      "foreground": "26b31a",
      "token": "constant.character.escape",
    },
    {
      "foreground": "26b31a",
      "token": "string source",
    },
    {
      "foreground": "0000ff",
      "token": "meta.preprocessor",
    },
    {
      "foreground": "0000ff",
      "token": "keyword.control.import",
    },
    {
      "foreground": "000000",
      "token": "entity.name.function",
    },
    {
      "foreground": "000000",
      "token": "keyword.other.name-of-parameter.objc",
    },
    {
      "foreground": "000000",
      "token": "entity.name.type",
    },
    {
      "foreground": "0000ff",
      "token": "storage.type",
    },
    {
      "foreground": "0000ff",
      "token": "storage.modifier",
    },
    {
      "foreground": "70727e",
      "token": "storage.type.method",
    },
    {
      "foreground": "000000",
      "token": "support.function",
    },
    {
      "foreground": "000000",
      "token": "support.class",
    },
    {
      "foreground": "000000",
      "token": "support.type",
    },
    {
      "foreground": "000000",
      "token": "support.constant",
    },
    {
      "foreground": "000000",
      "token": "support.variable",
    },
    {
      "foreground": "687687",
      "token": "keyword.operator.js",
    },
    {
      "background": "e1a09f",
      "token": "invalid",
    },
    {
      "background": "ffd0d0",
      "token": "invalid.deprecated.trailing-whitespace",
    },
    {
      "background": "427ff530",
      "token": "text source",
    },
    {
      "background": "427ff530",
      "token": "string.unquoted",
    },
    {
      "foreground": "68685b",
      "token": "meta.xml-processing",
    },
    {
      "foreground": "68685b",
      "token": "declaration.xml-processing",
    },
    {
      "foreground": "a31515",
      "token": "meta.doctype",
    },
    {
      "foreground": "a31515",
      "token": "declaration.doctype",
    },
    {
      "foreground": "0000ff",
      "token": "meta.tag",
    },
    {
      "foreground": "0000ff",
      "token": "declaration.tag",
    },
    {
      "foreground": "a31515",
      "token": "entity.name.tag",
    },
    {
      "foreground": "ff0000",
      "token": "entity.other.attribute-name",
    },
    {
      "foreground": "0000ff",
      "token": "string.quoted.double.xml",
    },
    {
      "foreground": "0000ff",
      "token": "string.quoted.double.html",
    },
    {
      "foreground": "0c07ff",
      "token": "markup.heading",
    },
    {
      "foreground": "000000",
      "token": "markup.quote",
    },
    {
      "foreground": "b90690",
      "token": "markup.list",
    },
  ],
}
`;
